fuuk/people/admin/options.py

```diff
@@ -147,39 +150,115 @@ class AgencyAdmin(TranslationAdmin):
-        if request.user.is_superuser:
-            return True
+        if obj:
+            if request.user == obj.editor or \
+                    request.user == getattr(obj.investigator_human, 'user', None):
+                return True
+            if obj.collaborators.filter(human__user=request.user).exists():
+                return True
+            return super(GrantAdmin, self).has_change_permission(request, obj)
+
+        return True
 
+    def has_delete_permission(self, request, obj=None):
+        """Returns True if the given request has permission to delete the given `Grant` instance
+
+        The delete permission has user, which is the editor, author or collaborator or has delete permission for grants.
+        If `obj` is None, this should return True if the given request has permission to delete *any* grant.
+        """
         if obj:
-            if request.user == getattr(obj.author.human, 'user', None):
+            if request.user == obj.editor or \
+                    request.user == getattr(obj.investigator_human, 'user', None):
                 return True
-            for author in obj.co_authors.all():
-                if request.user == getattr(author.human, 'user', None):
-                    return True
-            return False
+            if obj.collaborators.filter(human__user=request.user).exists():
+                return True
+            return super(GrantAdmin, self).has_delete_permission(request, obj)
 
-        return super(GrantAdmin, self).has_change_permission(request, obj)
+        return True
 
     def get_queryset(self, request):
+        """Returns a `QuerySet` of all `Grant` instances that can be edited by the admin site.
+
+        User with change permission can view all grants, other users can view only grants, where they are editor,
+        author or collaborator.
+        """
         queryset = super(GrantAdmin, self).get_queryset(request)
-        if request.user.is_superuser:
+        opts = self.opts
+        codename = get_permission_codename('change', opts)
+        if request.user.has_perm("%s.%s" % (opts.app_label, codename)):
             return queryset
-        human = request.user.human
-        return queryset.filter(
-            pk__in=queryset.filter(author__human=human) | queryset.filter(co_authors__human=human)
-        )
+        human = getattr(request.user, 'human', None)
+        if human:
+            return queryset.filter(Q(investigator_human=human) | Q(collaborators__human=human) | Q(editor=request.user))
+        else:
+            return queryset.filter(editor=request.user)
+
+    def save_model(self, request, obj, form, change):
+        """Given a model instance save it to the database.
+
+        The `editor` can be changed only by user with change permission to grants but `editor` field cannot be null.
+        Sets editor to the current user if it is a new `grant`.
+        """
+        if not change:
+            obj.editor = request.user
```

Comment:
Unexpected behavior: Grant admin will be set as an `editor` even though he specifically selected somebody else in the add form.